File: SinglyLinkedList.js
/**
 * Implements a singly linked list.
 * More info: https://leetcode.com/explore/learn/card/linked-list/209/singly-linked-list/1298/
*/

class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

class SinglyLinkedList {
    constructor() {
        this.size = 0;
        this.first = null;
        this.last = null;
    }

    /**
     * @returns {number}
     */
    getSize() {
        return this.size;
    }

    /**
     * @returns {boolean}
     */
    isEmpty() {
        return this.getSize() === 0;
    }

    /** Append a node of value val to the last element of the linked list. 
     * @param {number} value
     * @returns {void}
    */
    addLast(value) {
        let node = new Node(value);

        if (this.isEmpty()) {
            this.first = this.last = node;
        }
        else {
            this.last.next = node;
            this.last = node;
        }

        this.size++;
        return;
    }

    /** Add a node of value val before the first element of the linked list. 
     * After the insertion, the new node will be the first node of the linked list. 
     * @param {number} value
     * @returns {void}
    */
    addFirst(value) {
        let node = new Node(value);

        if (this.isEmpty()) {
            this.first = this.last = node;
        }
        else {
            node.next = this.first;
            this.first = node;
        }

        this.size++;
        return;
    }

    /** Add a node of value val before the index-th node in the linked list. 
     * If index equals to the length of linked list, the node will be appended 
     * to the end of linked list. If index is greater than the length, 
     * the node will not be inserted. 
     * @param {number} index
     * @param {number} value
     * @returns {void}
    */
    addAtIndex(index, value) {
        try {
            if (index < 0 || index > this.getSize()) {
                throw RangeError("Index must be between zero and linked list length (inclusive).");
            }
        } catch (error) {
            console.log(error);
            return;
        }
        if (index === 0) {
            this.addFirst(value);
        }
        if (index === this.getSize()) {
            this.addLast(value);
        }

        let current = new Node(value);
        let prev = this.getNode(index - 1);
        let next = prev.next;
        current.next = next;
        prev.next = current;

        this.size++;
        return;
    }

    /** Remove the first (i.e. zero-th) node in the linked list. 
     * If linked list is empty, method returns. 
     * @returns {number}
    */
    removeFirst() {
        try {
            if (this.isEmpty())
                throw Error("Linked list is empty.");
        } catch (error) {
            console.log(error);
            return;
        }

        let removedNodeValue = this.first.value;
        if (this.getSize() === 1) {
            this.first = this.last = null;
        }
        else {
            let swap = this.first.next;
            this.first.next = null;
            this.first = swap;
        }

        this.size--;
        return removedNodeValue;
    }

    /** Remove the last node in the linked list. 
     * If linked list is empty, method returns. 
     * @returns {number}
    */
    removeLast() {
        try {
            if (this.isEmpty())
                throw Error("Linked list is empty.");
        } catch (error) {
            console.log(error);
            return;
        }

        let removedNodeValue = this.last.value;
        if (this.getSize() === 1) {
            this.first = this.last = null;
        }
        else {
            let prev = this.getNode(this.getSize() - 2);
            this.last = prev;
            this.last.next = null;
        }

        this.size--;
        return removedNodeValue;
    }

    /** Remove the index-th node in the linked list. 
     * If linked list is empty or index is negative or index is greater than
     * linked list length, method returns. 
     * @param {number} index
     * @returns {number}
    */
    removeAtIndex(index) {
        try {
            if (this.isEmpty())
                throw Error("Linked list is empty.");
            if (index < 0 || index > this.getSize())
                throw RangeError("Index must be between zero and linked list length (inclusive).");
        } catch (error) {
            console.log(error);
            return;
        }

        if (this.isEmpty() || index < 0 || index > this.getSize() - 1) {
            return;
        }
        if (index === 0) {
            this.removeFirst();
        }
        if (index === this.size - 1) {
            this.removeLast();
        }

        let current = this.getNode(index);
        let previous = this.getNode(index - 1);
        let next = current.next;
        previous.next = next;

        this.size--;
        return current.value;

    }

    /** Helper function to return the index-th node in the linked list. 
     * @param {number} index
     * @returns {Node}
    */
    getNode(index) {
        let current = this.first;
        for (let i = 0; i < index && current !== null; i++) {
            current = current.next;
        }

        return current;
    }

    /** Get the value of the index-th node in the linked list. 
     * If the index is invalid, return -1. 
     * @param {number} index
     * @returns {number}
    */
    getNodeValue(index) {
        let current = this.getNode(index);
        return current === null ? -1 : current.value;
    }

    /** Add the value of each node in the linked list to an array, ordered from 
     * first to last. 
     * @returns {string}
    */
    toArray() {
        let array = [];
        let current = this.first;
        while (current !== null) {
            array.push(current.value);
            current = current.next;
        }

        return array;
    }
}

list = new SinglyLinkedList();
console.log(list.toArray());
console.log(list.getSize());
list.addLast(10);
list.addFirst(20);
console.log(list.toArray());
console.log(list.getSize());
list.addAtIndex(1, 30);
console.log(list.toArray());
console.log(list.getSize());
list.addAtIndex(2, 40);
console.log(list.toArray());
console.log(list.getSize());
console.log(list.removeAtIndex(2));
console.log(list.toArray());
console.log(list.getSize());
console.log(list.removeFirst());
console.log(list.toArray());
console.log(list.getSize());
console.log(list.removeLast());
console.log(list.toArray());
console.log(list.getSize());

File: singly_linked_list.py
# Implements a singly linked list.
# More info: https://leetcode.com/explore/learn/card/linked-list/209/singly-linked-list/1298/
from typing import List


class Node:
    def __init__(self, value: int) -> None:
        self.value = value
        self.next = None


class SinglyLinkedList:
    def __init__(self) -> None:
        self.size = 0
        self.first = None
        self.last = None

    def get_size(self) -> int:
        return self.size

    def is_empty(self) -> bool:
        return self.get_size() == 0

    def add_last(self, value: int) -> None:
        node = Node(value)
        if self.is_empty():
            self.first = self.last = node
        else:
            self.last.next = node
            self.last = node

        self.size += 1
        return

    def add_first(self, value: int) -> None:
        node = Node(value)
        if self.is_empty():
            self.first = self.last = node
        else:
            node.next = self.first
            self.first = node

        self.size += 1
        return

    def add_at_index(self, index: int, value: int) -> None:
        try:
            if self.is_empty() or index < 0 or index > self.get_size() - 1:
                raise IndexError()
        except IndexError:
            print(
                "Index Error: Index must be between zero and linked list length (inclusive).")
        else:
            if index == 0:
                self.add_first(value)
            if index == self.get_size():
                self.add_last(value)

            current = Node(value)
            previous = self.get_node(index - 1)
            next = previous.next
            current.next = next
            previous.next = current

            self.size += 1
            return

    def remove_last(self) -> int:
        try:
            if self.is_empty():
                raise RuntimeError()
        except RuntimeError:
            print("Error: Linked list is empty.")
        else:
            removed_node_value = self.last.value
            if self.get_size() == 1:
                self.first = self.last = None
            else:
                previous = self.get_node(self.get_size() - 2)
                self.last = previous
                self.last.next = None

            self.size -= 1
            return removed_node_value

    def remove_first(self) -> int:
        try:
            if self.is_empty():
                raise RuntimeError()
        except RuntimeError:
            print("Error: Linked list is empty.")
        else:
            removed_node_value = self.first.value
            if self.get_size() == 1:
                self.first = self.last = None
            else:
                swap = self.first.next
                self.first.next = None
                self.first = swap

            self.size -= 1
            return removed_node_value

    def remove_at_index(self, index: int) -> int:
        try:
            if self.is_empty():
                raise RuntimeError()
            if index < 0 or index > self.get_size() - 1:
                raise IndexError()
        except RuntimeError:
            print("Error: Linked list is empty.")
        except IndexError:
            print(
                "Index Error: Index must be between zero and linked list length (inclusive).")
        else:
            if index == 0:
                self.remove_first()
            if index == self.get_size() - 1:
                self.remove_last()

            current = self.get_node(index)
            previous = self.get_node(index - 1)
            next = current.next
            previous.next = next

            self.size -= 1
            return current.value

    def get_node(self, index: int) -> Node:
        current = self.first
        for i in range(index):
            if current is not None:
                current = current.next
            else:
                break

        return current

    def get_node_value(self, index: int) -> int:
        current = self.get_node(index)
        if current is None:
            return -1
        else:
            return current.value

    def to_array(self) -> List[int]:
        array = []
        current = self.first
        while current is not None:
            array.append(current.value)
            current = current.next

        return array


my_list = SinglyLinkedList()
print(my_list.to_array())
print(my_list.get_size())
my_list.add_last(10)
my_list.add_first(20)
print(my_list.to_array())
print(my_list.get_size())
my_list.add_at_index(1, 30)
print(my_list.to_array())
print(my_list.get_size())
my_list.add_at_index(2, 40)
print(my_list.to_array())
print(my_list.get_size())
print(my_list.remove_at_index(2))
print(my_list.to_array())
print(my_list.get_size())
print(my_list.remove_first())
print(my_list.to_array())
print(my_list.get_size())
print(my_list.remove_last())
print(my_list.to_array())
print(my_list.get_size())
